Sources/ProjectDescription/FileElement.swift

```diff
@@ -9,11 +9,11 @@ import Foundation
 ///       `"some/pattern/**"` is the equivalent of `FileElement.glob(pattern: "some/pattern/**")`
 public enum FileElement: Codable, Equatable {
     /// A glob pattern of files to include
-    case glob(pattern: Path)
+    case glob(pattern: Path, tags: [String] = [])
 
     /// Relative path to a directory to include
     /// as a folder reference
-    case folderReference(path: Path)
+    case folderReference(path: Path, tags: [String] = [])
```

Comment:
+1 on the backwards compatibility, that said extending `FileElement` to support tags will introduce some chances for error.

Today `FileElement` is used in a few places:

- `Target.resources`
- `Project.additionalFiles`
- `Workspace.additionalFiles`

Tags are only applicable to the first use case and having it the other cases may add some API confusion:

e.g.
```swift
let project = Project(
    // ...
   additionalFiles: [
        .folderReference(path: "Documentation", tags: ["Docs"]) // tags here have no effect
   ]
)
```

As such I propose we introduce a new type `ResourceFileElement` which has the additional tags property and is only used for `Target.resources` to help alleviate the incorrect cases.

Reply:
thank you @kwridan @thedavidharris!
Will add `ResourceFileElement` for resources.

Reply:
So I tried to add `ResourceFileElement`, but faced an issue. 

Right now Tuist have this `Array` extension
``` swift
extension Array: ExpressibleByStringLiteral where Element == FileElement {
    public typealias StringLiteralType = String

    public init(stringLiteral value: String) {
        self = [.glob(pattern: Path(value))]
    }
}

// so users can
let resources: [FileElement] = "/path"
```

But we can't apply same to any other type, because [overlapping conformances are not allowed in Swift](https://github.com/apple/swift-evolution/blob/master/proposals/0143-conditional-conformances.md#overlapping-conformances).

We can make a workaround like this:
``` swift
public protocol StringLiteralInitializable: ExpressibleByStringLiteral {}

extension StringLiteralInitializable where Self == [ResourceFileElement] {
    public init(stringLiteral value: String) {
        self = [.glob(pattern: Path(value))]
    }
}

extension StringLiteralInitializable where Self == [FileElement] {
    public init(stringLiteral value: String) {
        self = [.glob(pattern: Path(value))]
    }
}

extension Array: StringLiteralInitializable {
    public init(stringLiteral value: String) {
        self = []
    }
}
```
But in this case it would be possible to initialize any array with `String` literal.
``` swift
let array: [Int] = "" // []
```

Another option is to remove those extensions at all and add additional initializers to `Project`, `Workspace` and `Target`. But I personally don't really like this option.

@kwridan wdyt? Should I go with workaround?

Reply:
Ah good catch - we could solve this by introducing another type `ResourceFileElements` (with an `s`) that is wrapper to `[ResourceFileElement]`, the benefit we get here is we can conform `ResourceFileElements` to both `ExpressibleByStringLiteral` and `ExpressibleByArrayLiteral` to maintain backwards compatibility. 

Would that work?

Reply:
Yeah, it works. Great idea!